### ISSUE TITLE
Spaces inserted when terminating a statement with hard_tabs on

### ISSUE THREAD

**Issue by levvij** [zed-industries/zed#13733]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

set hard_tabs to true, create a new file `test.ts`: 
```
const x = [
<TAB>1234, 123
]
```

Zed inserts and shows a tab.
Add a `;` after `]`. Zed will automatically remove the Tab and replace it with 4 spaces.

other options in my settings that might have an effect on this:
"show_whitespaces": "boundary"
"hard_tabs": true
"tab_size": 4
"format_on_save": "off"
"remove_trailing_whitespace_on_save": false

### Environment

Zed: v0.141.3 (Zed)
OS: macOS 14.5.0
Memory: 32 GiB
Architecture: aarch64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by notpeter** [comment]
I can reproduce this issue in preview with your settings:
```
  "preview": {
    "languages": {
      "TypeScript": {
        "show_whitespaces": "boundary",
        "hard_tabs": true,
        "tab_size": 4,
        "format_on_save": "off",
        "remove_trailing_whitespace_on_save": false
      }
    }
  },
```
https://github.com/zed-industries/zed/assets/145113/2d789dcf-0817-4a3a-b173-3a6d21d77ae0

**Comment by goldenice** [comment]
Just adding a data point for me: it appears setting `"use_on_type_format": false` stops the behavior from occurring.

I also tried the following things, which had no effect:
- Added hard_tabs: true specifically for languages JavaScript/TypeScript/TSX
- Added hard_tabs: true in the project settings.json instead of my global settings.json
- Tried to configure ESLint --fix to autoformat as a code_actions_on_format to see if that would get rid of the wrongly inserted spaces (it didn't, although I struggled to get ESLint to autoformat other things too. Pretty sure I saw it applying a different rule once though.)
- Enabled/disabled format_on_save
- Changed tab_size
- Enabled/disabled remove_trailing_whitespace_on_save
- Tried all settings of show_whitespaces
- Enabled/disabled show_inline_completions

**Comment by jakubhruby** [comment]
I have the same issue. I'm not able downgrade to investigate it, but I think it has broken in version 0.141.x.

**Comment by matubu** [comment]
When you disable the vtsls language server, the issue disappears.

```json
"languages": {
	"JavaScript": {
		"language_servers": ["!vtsls"]
	}
}
```

It seems to be related to [this option](https://github.com/yioneko/vtsls/blob/c53d716ad44a527e21372e93fe7f2f894d2e03dd/packages/service/configuration.schema.json#L293C6-L293C64), although I don't know how to change it.

**Comment by levvij** [comment]
I found the issue:
 
setting this will fix it:
 "use_on_type_format": false

for future updates that add new functions like this, it might be helpful if the new options would automatically be configured in the user config, while defaulting to something that would match the old behaviour. 

maybe, "use_on_type_format": false would have automatically added to the config, if the user had "format_on_save" enabled before the update. because "format_on_save" inheriting to "use_on_type_format" might not be everyones preference.

**Comment by matubu** [comment]
This should now be fixed by #14536. And insert hard tabs instead if your config says so.
